### ISSUE TITLE
Esmini is unable to control the obstacles stop conditions

### ISSUE THREAD

**Issue by Iqbalparvi** [esmini/esmini#320]
Hello;
We are experimenting by simulating the ADAS-related driving scenarios with the Esmini.
It seems that the Esmini simulator cannot control the obstacle to follow the predefine routing actions in the Open scenario.
The yellow line in the below figure shows our defined waypoints/trajectories to follow the ego car and the target car. At the start of the simulation, everything is performed as expected. However, the system ignores the ending points of both obstacles to stop the simulation.
![screen_shot_Frame00260bata564](https://user-images.githubusercontent.com/47267894/190314253-b406cd35-9c1c-4593-b517-c8280594339c.png)
Can you please investigate why the system behavior differs from the design view, whether it is any limitation, a bug in esmini, or something else?
Regards

**Comment by eknabevcc** [comment]
Please provide the scenario (or another with the same issue) for our understanding and investigation.

**Comment by Iqbalparvi** [comment]
Hi @eknabevcc ;
Thanks for your prompt response.
It’s my bad I forget to attach the scenario.
Please find it here:
[Scenario.zip](https://github.com/esmini/esmini/files/9579533/Scenario.zip)
Further that our explanation, the stop positions for both VUT and GVT are
`<Position><WorldPosition x="150" y="15" z="0"/></Position>) `
and 
`(<Position> <WorldPosition x="118.23" y="0" z="0"/> </Position>`
 respectively.
It should be noted that the GVT (Red car) final waypoint point is smaller than that of the VUT (white car) which is considered its destination where it must be stopped.
Regards.

**Comment by eknabevcc** [comment]
Thanks for the scenario!

I first thought that you expected the cars to stop at the end of trajectories. But now when I see the scenario and your additional explanation I (think that I) realize that you make use of ReachPositionConditions to trig the end of the scenario. I think that's a feasible solution. 

Looking at the scenario, at least with on my side, both cars travels along y=0. So changing first stop condition from:
`<Position><WorldPosition x="150" y="15" z="0"/></Position>)`
to:
`<Position><WorldPosition x="150" y="0" z="0"/></Position>)`
I think it stops as expected. Can you try and see if you get same behavior?

**Comment by Iqbalparvi** [comment]
Thank you so much for the follow-ups.
I have checked and changed the y condition to 0 for the Ego car to stop, and yes, Its stops at the last waypoints.
That is 
`<Position><WorldPosition x="150" y="0" z="0"/></Position>`
However, the GVT vehicle (red car) is still unable to stop at its last waypoints i.e 
`<WorldPosition x="118.23" y="0" z="0"/>.`
This vehicle should stop earlier as its **ReachPositionCondition** is smaller than another vehicle.
The below figure compares the scenario design interface and the Esmini Closed loop simulation windows.
![INputVSOutput](https://user-images.githubusercontent.com/47267894/190835032-b4e86987-1deb-4e15-958c-fea2c141d063.png)
In the design phase, the blue car (i.e., the Ego car) follows the blue dots and lines, while the red car is GVT and follows the red line and dots. In this phase, at the time (T=5.8s), the red car stops at its destination while the blue car follows its waypoint and moves towards the destination to stop at T=9.2s.
When feeding the scenario to achieve the same behaviors on Esmini, the red car cannot stop at the expected position (x=118.23) at T=5.8s and continue moving ahead. The simulation stops when the ego (white) car reached its destination.
To understand more please see the two videos as well.
[Scenario_mod.zip](https://github.com/esmini/esmini/files/9590685/Scenario_mod.zip)
[vedios.zip](https://github.com/esmini/esmini/files/9590686/vedios.zip)

**Comment by eknabevcc** [comment]
Ah, I think I see it. You have an empty [TimeReference](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/TimeReference.html) element. 

It can be argued, but esmini will interpret missing TimeReference as TimeReference/None.

Try change the two occurrences of:
```xml
<TimeReference/>
```
into:
```xml
<TimeReference>
     <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/>
</TimeReference>
```

Then I see behavior much closer to what you have in the editor. 

BTW: Can you tell what editor is this? Looks nice.

**Comment by Iqbalparvi** [comment]
The finding is useful, and thanks for showing interest in the scenario design interphase, which is a part of the MATLAB/Simulink Automated driving toolbox [where](https://au.mathworks.com/products/automated-driving.html) the ADAS Simulink system performs as expected.
By following your above recommendation, we have now got the expected results on our lab screen in the Esmini OSC window.
However, at this stage,

- **can we assume that Esmini always ignores and unused the controlling behaviors to Trig the ReachpositionConditions, or it may be something like dynamic limitations in the scenarioReader.cpp?.

- And in the future, is it achievable on Esmini with TimeReference/None?

**
Regards.

**Comment by eknabevcc** [comment]
**Q1: Can we assume that Esmini always ignores and unused the controlling behaviors to Trig the ReachpositionConditions, or it may be something like dynamic limitations in the scenarioReader.cpp?**

Maybe a misunderstanding? esmini do **not** ignore the stop conditions. The stop condition was wrong in your first file. As I pointed out, the car does never reach near y=15, that's the reason why the condition is not becoming `true` (and since you have two separate conditionGroups, both must be true (AND)). If you change first condition to y=0 it works as expected -> the scenario stops due to stopTriggers. 
Or do I misunderstand you question?

That said, it's a bit tricky with combined ReachPosition conditions if cars are moving and just passing positions. As soon as one car is moving away from the position the condition is not true anymore. Maybe [TraveledDistanceCondition](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/TraveledDistanceCondition.html) is a better choice since it will return `true` continuously after reaching and passing specified distance.

**Q2: And in the future, is it achievable on Esmini with TimeReference/None?**
The control points of a trajectory may include timeStamps which then can be used to control the speed of the entity (timeRef relative/absolute). However, for various reasons it can be useful to ignore the timestamps and set instead set speed with other actions, e.g. `speedAction`, while entity is still moving along the trajectory. Instead of having to remove all timestamps (`time` attribute), you can simply set timeReference/None. 

Now, what if the timeReference is empty? esmini interpret this as timeReference/None, based on this statement:

`Defines if time information provided within the trajectory should be considered. If so, it may be used as either absolute or relative time along the trajectory in order to define longitudinal velocity of the actor. Moreover, a time offset or time scaling may be applied.` 
in [FollowTrajectoryAction](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/FollowTrajectoryAction.html)

So, by either setting timeReference/None or leaving it empty esmini will ignore timestamps. And by defining timeReference/Timing element, to either Relative or Absolute, the timestamps will have effect.

I think it makes sense to keep it this way. Otherwise the standard needs to be changed in terms of default timeReference behavior in case timeReference is empty to be either Timing = Relative or Timing = Absolute.

**Comment by Iqbalparvi** [comment]
All right, Agreed with your 2nd suggestion. However, I think the value y=15 was a typo mistake in my 1st file during a modification in the open Scenario.
When I compare it to my scenario design function conditionGroup, it was actually y=0, as in the below figure.
![fn](https://user-images.githubusercontent.com/47267894/190961181-2d2a4ec8-e31e-4ffc-a334-b9d743336ae7.PNG)
I will reopen the discussion later on if feel necessary.
Anyways, thanks again for your guidance and fruitful discussion.
Regards.

**Comment by eknabevcc** [comment]
Ok, I see. Let's forget about the y=15 typo. But then I don't understand your first question: 
**Q1: Can we assume that Esmini always ignores and unused the controlling behaviors to Trig the ReachpositionConditions, or it may be something like dynamic limitations in the scenarioReader.cpp?**

What I can see esmini DO trigger on the stop conditions (when typo is fixed). What is the issue that you see?

**Comment by Iqbalparvi** [comment]
Ok, as we already provide the two videos to differentiate the design level simulation and the Esmini output view. 
Let me shed light again thoroughly.
1. When the typo is fixed, we can see that the VUT i.e white car stops at the end of the yellow line which is its StopCondition. However, the red car cannot stop at the expected position (x=118.23, 0, 0) at T=5.8s and continues moving ahead.
Please see the attached scenario and video1 again.
[scenario1&vedio1.zip](https://github.com/esmini/esmini/files/9603188/scenario1.vedio1.zip)

2. To achieve the Red vehicle behavior, we assumed and found that this is due to the time reference/None in the scenario i..e <TimeReference/>, which esmini will interpret missing TimeReference.

By following your recommendations, when we changed the two occurrences into 
`<TimeReference>
     <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/>
</TimeReference>`
We see that expected behaviors as the GVT (red car) stop earlier than that of VUT.
Please see the attached scenario and video2 again.
[Scenario2&vedio2.zip](https://github.com/esmini/esmini/files/9603192/Scenario2.vedio2.zip)

So, we raised this question **Q1** based on the above two different situations, as we are not sure about the 1st condition, where Esmini really following or not the define trajectories for the Ego car to stop at the EndConditionPosition i.e 
`<Position><WorldPosition x="150" y="0" z="0"/></Position>)`
 with <TimeReference/>.
When, esmini needs a TimeReference value to stop the target car at the desired position, i.e., earlier than the Ego car.
It also assumes and expects that to achieve the goal, the performance should be the same at different simulators/platforms for the same scenario without changing further conditions.

**Comment by eknabevcc** [comment]
Thanks for additional info. I think we have an interesting case which is worth to try sort out. But to simplify analysis, let's focus on one scenario at a time. Starting with your first scenario, this is my conclusions:

There are two stop conditions. They are in separated groups which means that the trigger will happen as soon as at least one of them become TRUE (OR). Previously I wrongly stated that they both needed to be TRUE (AND), but I was wrong - my bad. Conditions within same [group](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/ConditionGroup.html) are AND:ed while groups in a [trigger](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/Trigger.html) are OR:ed (I'll try to remember this for ever 🙂). Also the second condition has a delay of 6 seconds.

After quick analysis focusing on stop trigger conditions, this is what I find happens:
t=5.7: Target reaches 2m radius from position x=118.23, y=0. Timer of 6 seconds started - no trigger yet.
t=9.2: Ego reaches 2m radius from x=150, y=0. No delay => this condition trigger scenario stop.

If this is not according to your understanding, please describe what is your expectations of scenario 1?

**Comment by Iqbalparvi** [comment]
**If this is not according to your understanding, please describe what is your expectations of scenario 1?**
Thank you for your understanding and valuable response where your observation is right as:
**Condition-I**
 At t=5.7: Target reaches 2m radius from position x=118.23, y=0. Timer of 6 seconds started - no trigger yet. 
But also, it should be noted that the target car is not stop here.
And at 
 **Condition -II**
 t=9.2: Ego reaches 2m radius from x=150, y=0. No delay => this condition trigger scenario stops, also the Ego car stops here.
From this scenario1, our expectation is that when the target is reached at its position i.e Condition-I, it should stop. However, the simulation should continue until the Ego car reaches its position which is condition-II, and then triggers the scenario.
In short, the behavior that we have observed after modification in the scenario1 with the <TimeRefernce> value that is in scenario2 and vedio2 is our expected output from the main scenario.
**Note:**
We are not sure whether it is a limitation or a shortcoming with the Esmini simulator or with the driving scenario designer platforms.
In both case we didn’t know much more about the internal core idea of the algorithms, that's why expects that if the driving scenario designer on other simulator/platforms have the option to transform (import/export) the scenario into OpenX format,  then the relevant scenario players/simulators like Esmini, Carla, etc must perform the desired actions when simulating the scenarios accordingly.
Further, we have run this scenario with a few other open and closed simulators/platforms where the expected results are achieved.

**Comment by eknabevcc** [comment]
I agree we should try sort it out to have a common behavior cross simulators 👍

Let's focus on the first deviating expectation: Scenario 1 and condition 1 (EndCondition_Global Vehicle Target_2): You expect the Target car to stop immediately when condition is true, at t=5.7, while I do not.

My understanding: The condition is part the Act stopTrigger, which ONLY effect is to stop the complete act. Target car do trigger the condition, but is not affected by the condition in any way. Further, the delay is postponing the effect of the trigger, i.e. "Stop the act, but first wait 6 seconds". So my expectation is that Target should move on and Act should be stopped after 6 seconds (or earlier if other triggers happens meanwhile).

Can you clarify, what makes you expect the Target to stop at condition 1, at 5.7 seconds? 

Maybe you have a different understanding of the `TriggeringEntities` element which contains the Target. As stated above, it only means that Target is monitored for reaching the position and then triggering the condition. The stopTrigger do not apply to Target specifically, but the complete Act.

**Comment by Iqbalparvi** [comment]
** Can you clarify, what makes you expect the Target to stop at condition 1, at 5.7 seconds?**
Sorry for the confusion, Let’s have a look at the below figure; it shows my use case overview in the scenario design phase.
![1](https://user-images.githubusercontent.com/47267894/191873700-ebffaa54-d9fe-40a8-bb8c-c355df83e702.PNG)
Just for clarification, I have split the two trajectories that each vehicle needs to follow to travel from its initial position to its final position (stop condition).
The Blue car has its starting and stopping waypoints, and the red car has its own. The red line shows that the target car waypoint/Trajectory starts at position x=36.7 and ends at x=118.23.
`<TrajectoryFollowingMode followingMode="position"/>`
For further observation of the expected behavior, please see the video again.
https://user-images.githubusercontent.com/47267894/191873983-eafcccb5-97f1-4f0a-8055-53c644500a90.mp4
In summary, when there are no trajectories to follow for the target car after x=118.23, then how can we expect it to move on?
It makes sense and I agree with you that the complete scenario stopTrigger applies to the complete Act until all the acts are not done.

**Comment by eknabevcc** [comment]
Ah, now I see the point of different expectations: What should happen when car reach end of trajectory?

I see use cases where it makes sense to stop at end of a trajectory, for example the car is colliding or braking to a stop. 
But I also see a lot of use cases where the car should just continue at current speed, for example when using a trajectory to define the shape of a lane change maneuver. 

So, to keep door open for all use cases esmini default behavior, after end of trajectory, is to hand over to default controller to move the car along the road with current speed and lane offset. 

If you want to stop the car at end of trajectory there are two simple ways (I think your intention is to use a combination of both):

1. Control speed by time stamps
   - To have a complete stop at end, create a final control point based on the last one: Same position but at later timestamp, e.g. one second later. This will make the car stop. 
   - Example: [Scenario1-mod1.xosc.txt](https://github.com/esmini/esmini/files/9631269/Scenario1-mod1.xosc.txt)

2. Add a [SpeedAction](https://www.asam.net/static_downloads/ASAM_OpenSCENARIO_V1.2.0_Model_Documentation/modelDocumentation/content/SpeedAction.html) to stop the car.
   - Trig action at end of FollowTrajectory action
   - Set step shape for immediate stop, or use rate for ramping down to zero.
   - Example (step): [Scenario1-mod2.xosc.txt](https://github.com/esmini/esmini/files/9631275/Scenario1-mod2.xosc.txt)
  
   To enable use of `StoryboardElementStateCondition` I moved the FollowTrajectory actions from Init to Story maneuvers.
   I also modified Act stop conditions to wait for both maneuverGroups has finished (put the two conditions in one group - AND)

**Comment by Iqbalparvi** [comment]
Thank you for all your time and efforts to see the use case from different angles.
I agree with you that there are different use cases where the target car is moving with current speed, applying a brake, stationery, etc at the end of the trajectory.
For example, one of our use case in another simulator says **A car-to-car rear moving (CCRm) scenario, where the ego vehicle rear-ends a moving vehicle. At collision time, the ego vehicle overlaps with 50% of the width of the moving vehicle.**

To see a common behavior across simulators, this exploration is interesting and helpful in terms of scenario catalogs and tools/system design mechanisms.
Although the expectation is achieved with the suggested modifications, however, should we conclude the case in such a way if you agree with me?

- To run this type of TrajectoryFollowing scenario in terms of OpenX on Esmini the value to <TimeReference> and TimeStamp is important.

`<TimeReference>
                    <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/>
                </TimeReference>`
If setting timeReference/None or leaving it empty esmini will ignore timestamps.

- The exploration helps us to understand OpenX scenario behaviors and also helps us to include more scenarios in the catalog of Esmini.

**Comment by eknabevcc** [comment]
Yes, I totally agree with you:

`<TimeReference> <Timing domainAbsoluteRelative="relative" offset="0" scale="1"/> </TimeReference>` will use timestamps relative the start time of the action. While `domainAbsoluteRelative="absolute"` would use timestamps as is, relative simulation time 0 (start of scenario) - which may be useful in some cases.

And if TimeReference is missing or set to <None> then timestamps will be ignored.

Thanks for this discussion which aims to harmonize the behavior of different tools 👍

**Comment by Iqbalparvi** [comment]
Thank you so much for your time and efforts. I am going to close the discussion for now.
A great pleasure throughout the discussion in this amazing environment.
We will continue our exploration in the future and thanks to the whole Esmini team for providing such an easy and understandable environment.
Best Regards;
